proposals/devcontainer-templates.md

```diff
@@ -0,0 +1,192 @@
+  4. See the Template's `README` for configuration options.
+  5. Run **Remote-Containers: Reopen in Container** to use it locally, or **Codespaces: Rebuild Container** from within a codespace.
+
+### Referencing a Template 
+
+The `id` format (`<oci-registry>/<namespace>/<template>[:latest]`) dictates how a [supporting tool](#template-supporting-tools-and-services) will locate and download a given Template from an OCI registry (For example - `ghcr.io/user/repo/go:latest`). The registry must implement the [OCI Artifact Distribution Specification](https://github.com/opencontainers/distribution-spec). Some implementors can be [found here](https://oras.land/implementors/). 
+
+## Versioning
+
+Each Template is published with only the `latest` tag. Tooling that handles releasing Templates will republish the `latest` tag every time a new release is created.
+
+## Release
+
+_For information on distribution Templates, see [devcontainer-templates-distribution.md](./devcontainer-templates-distribution.md)._
+
+### Option Resolution
+
+A templates's `options` is used by a supporting tool to prompt for different configuration options. A supporting tool will parse the `options` object provided by the user. If a value is selected for a Template, it will be replaced in the provided `replaceIn` files.
+
+### Option resolution example
+
+Consider a `java` template with the following folder structure:
+
++-- java
+|    +-- devcontainer-template.json
+|    +-- .devcontainer.json
+
+Suppose the `java` Template has the following `options` parameters declared in the `devcontainer-template.json` file:
+
+```json
+// ...
+"options": {
+    "imageVariant": {
+        "type": "string",
+        "description": "Specify version of java.",
+        "proposals": [
+          "17-bullseye",
+          "17-buster",
+          "11-bullseye",
+          "11-buster",
+          "17",
+          "11"
+        ],
+			  "default": "17-bullseye",
+        "replaceIn": [".devcontainer.json"]
+    },
+    "nodeVersion": {
+        "type": "string", 
+        "proposals": [
+          "latest",
+          "16",
+          "14",
+          "10",
+          "none"
+        ],
+        "default": "latest",
+        "description": "Specify version of node, or 'none' to skip node installation.",
+        "replaceIn": [".devcontainer.json"]
+    },
+    "installMaven": {
+        "type": "boolean", 
+        "description": "Install Maven, a management tool for Java.",
+        "default": "false",
+        "replaceIn": [".devcontainer.json"]
+    },
+}
+```
+
+and it has the following `.devcontainer.json` file:
+
+```json
+{
+	"name": "Java",
+	"image": "mcr.microsoft.com/devcontainers/java:0-{{imageVariant}}",
+	"features": {
+		"ghcr.io/devcontainers/features/node:1": {
+			"version": "{{nodeVersion}}",
```

Comment:
We are using `${localEnv:<env var name>}` and similar variables at runtime. Should we use `${templateOption:<option id>}` (e.g., `${templateOption:nodeVersion}` here) for consistency?

Reply:
Nice, updated!